compiler/src/lnntoamm/Stmt.ts

```diff
@@ -0,0 +1,351 @@
+import { LPNode } from '../lp';
+import Output from './Amm';
+import Event from './Event';
+import Expr, { Ref } from './Expr';
+import opcodes from './opcodes';
+import Scope from './Scope';
+import Type from './Types';
+import { genName, TODO } from './util';
+
+export class MetaData {
+  scope: Scope
+  variables: { [name: string]: VarDef }
+  retTy: Type
+
+  constructor(
+    scope: Scope,
+    retTy: Type,
+    variables: { [name: string]: VarDef } = null,
+  ) {
+    this.scope = scope;
+    this.retTy = retTy;
+    this.variables = variables !== null ? variables : {};
+  }
+
+  get(name: string): VarDef | null {
+    if (!this.variables.hasOwnProperty(name)) {
+      return null;
+    }
+    return this.variables[name];
+  }
+
+  define(dec: VarDef) {
+    if (this.get(dec.name) !== null) {
+      throw new Error(`Can't redefine value ${dec.name}`);
+    }
+    this.variables[dec.name] = dec;
+  }
+}
+
+export default abstract class Stmt {
+  ast: LPNode
+
+  constructor(
+    ast: LPNode,
+  ) {
+    this.ast = ast;
+  }
+
+  abstract inline(amm: Output): void;
+
+  static fromAst(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts = [];
+    if (ast.has('assignables')) {
+      let [ss, expr] = Expr.fromAssignablesAst(ast.get('assignments'), metadata);
+      stmts.push(...ss, Dec.gen(expr, metadata));
+    } else if (ast.has('assignments')) {
+      stmts.push(...Assign.fromAssignments(ast.get('assignments'), metadata));
+    } else if (ast.has('conditionals')) {
+      stmts.push(...Cond.fromConditionals(ast.get('conditionals'), metadata));
+    } else if (ast.has('declarations')) {
+      stmts.push(...Dec.fromDeclarations(ast.get('declarations'), metadata));
+    } else if (ast.has('emits')) {
+      stmts.push(...Emit.fromEmits(ast.get('emits'), metadata));
+    } else if (ast.has('exits')) {
+      stmts.push(...Exit.fromExits(ast.get('exits'), metadata));
+    } else {
+      throw new Error(`unrecognized statement ast: ${ast}`);
+    }
+    return stmts;
+  }
+}
+
+class Assign extends Stmt {
+  upstream: VarDef
+  expr: Expr
+
+  constructor(
+    ast: LPNode,
+    upstream: VarDef,
+    expr: Expr,
+  ) {
+    super(ast);
+    this.upstream = upstream;
+    this.expr = expr;
+    if (this.upstream.immutable) {
+      throw new Error(`cannot reassign to a const value`);
```

Comment:
Nit: maybe include the name of the variable in this error message?

Though that can be a future task to add better error messages everywhere, ideally including file names and line numbers.

Reply:
I do intend for there to be a future task focused on improving error messages (eg, actually using the `ast` kept by all of the nodes), but this is simple enough